File: python/train.py
import numpy as np
import torch
import torch.nn.functional as F
from nerf_model import NeRF
from constants import RESULT_DIR
import os
import time
import argparse

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--max_step", type=int, default=1e9)
    args = parser.parse_args()

    dataset = np.load(f'{RESULT_DIR}/dataset.npz')
    o_np = dataset['o']
    d_np = dataset['d']
    C_np = dataset['C']

    N_EPOCH = 1
    BATCH_SIZE = 2048
    MAX_STEP = args.max_step
    PRINT_INTERVAL = 100

    nerf = NeRF(t_n=0., t_f=2.5)
    nerf.to("cuda")

    optimizer = torch.optim.Adam(
        nerf.parameters(),
        lr=2e-4, betas=(0.9, 0.999), eps=1e-7)

    n_sample = o_np.shape[0]

    save_dir = f"{RESULT_DIR}/train/"
    os.makedirs(save_dir, exist_ok=True)

    f = open(f"{save_dir}/train_loss.tsv", 'w')
    header_str = "time\tepoch\tstep\tepoch_rate\tloss\n"
    f.write(header_str)
    print(header_str, end="")

    step = 0
    start_time = time.time()

    for e in range(1, N_EPOCH + 1):
        perm = np.random.permutation(n_sample)
        sum_loss = 0.
        sample_num = 0

        for i in range(0, n_sample, BATCH_SIZE):
            step += 1
            o = o_np[perm[i:i + BATCH_SIZE]]
            d = d_np[perm[i:i + BATCH_SIZE]]
            C = C_np[perm[i:i + BATCH_SIZE]]
            o = torch.tensor(o, device=nerf.device())
            d = torch.tensor(d, device=nerf.device())
            C = torch.tensor(C, device=nerf.device())

            C_c, C_f = nerf.forward(o, d)
            loss = F.mse_loss(C_c, C) + F.mse_loss(C_f, C)
            sum_loss += loss.item() * o.shape[0]
            sample_num += o.shape[0]

            if (i / BATCH_SIZE + 1) % PRINT_INTERVAL == 0:
                sum_loss /= sample_num
                elapsed_sec = int(time.time() - start_time)
                elapsed_min = elapsed_sec // 60 % 60
                elapsed_hou = elapsed_sec // 3600
                elapsed_str = f"{elapsed_hou:02d}:{elapsed_min:02d}:{elapsed_sec % 60:02d}"
                epoch_rate = 100 * (i + BATCH_SIZE) / n_sample
                print_str = f"{elapsed_str}\t{e}\t{step}\t{epoch_rate:5.1f}\t{sum_loss:.5f}\n"
                f.write(print_str)
                f.flush()
                print(print_str, end="")
                sum_loss = 0
                sample_num = 0
                # save state.
                torch.save(nerf.state_dict(), f"{save_dir}/nerf_model.pt")
                torch.save(optimizer.state_dict(), f"{save_dir}/optimizer.pt")

            optimizer.zero_grad()
            loss.backward()  # type: ignore
            optimizer.step()
            if step >= MAX_STEP:
                break

        if step >= MAX_STEP:
            break

File: cpp/main/train.cpp
#include <filesystem>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <random>

#include "../data.hpp"
#include "../nerf_function.hpp"
#include "../nerf_model.hpp"
#include "../timer.hpp"
#include "../utils.hpp"

int main() {
  const std::string dataset_path = "../data/train/greek/";

  const CameraIntrinsicParameter param = GetCameraIntrinsicParameter(dataset_path);

  const std::vector<std::string> pose_paths = Glob(dataset_path + "pose/");
  const std::vector<std::string> image_paths = Glob(dataset_path + "rgb/");
  assert(pose_paths.size() == image_paths.size());
  const int32_t N = pose_paths.size();
  std::vector<Data> dataset_raw(N);
  std::vector<RayData> ray_data;
  for (int32_t i = 0; i < N; i++) {
    std::cout << pose_paths[i] << " " << image_paths[i] << std::endl;
    dataset_raw[i].pose = ParsePose(pose_paths[i]);
    dataset_raw[i].image = cv::imread(image_paths[i]);
    std::vector<RayData> curr_rays = GetRays(param, dataset_raw[i]);
    ray_data.insert(ray_data.end(), curr_rays.begin(), curr_rays.end());
  }

  constexpr int32_t kEpoch = 1;
  constexpr int32_t kBatchSize = 2048;
  constexpr int32_t kMaxStep = 1000;
  constexpr int32_t kPrintInterval = 100;

  NeRF nerf;
  nerf->to(torch::kCUDA);
  torch::optim::AdamOptions options;
  options.lr(3e-4);
  options.betas({0.9, 0.999});
  options.eps(1e-7);
  torch::optim::Adam optimizer(nerf->parameters(), options);

  const int32_t n_sample = ray_data.size();
  std::cout << "n_sample = " << n_sample << std::endl;

  const std::string result_dir = "result_dir/";
  const std::string save_dir = result_dir + "train/";
  std::filesystem::create_directories(save_dir);
  std::ofstream ofs(save_dir + "train_loss.tsv");
  const std::string header_str = "time\tepoch\tstep\tepoch_rate\tloss\n";
  ofs << header_str;
  std::cout << header_str;

  int32_t step = 0;
  Timer timer;
  timer.Start();

  std::mt19937_64 engine(std::random_device{}());
  for (int32_t e = 1; e <= kEpoch; e++) {
    std::shuffle(ray_data.begin(), ray_data.end(), engine);
    float sum_loss = 0.0f;
    int64_t sample_num = 0;

    for (int32_t i = 0; i < n_sample; i += kBatchSize) {
      step++;

      // ミニバッチを作成
      const std::vector<RayData> curr_data(ray_data.begin() + i, ray_data.begin() + std::min(i + kBatchSize, n_sample));
      auto [o, d, C] = RayData2Tensor(curr_data);
      C = C.to(nerf->device());
      C /= 255;

      // 推論
      auto [C_c, C_f] = nerf->forward(o, d);
      torch::Tensor loss = torch::nn::functional::mse_loss(C_c, C) + torch::nn::functional::mse_loss(C_f, C);
      sum_loss += loss.item<float>() * o.size(0);
      sample_num += o.size(0);

      if ((i / kBatchSize + 1) % kPrintInterval == 0) {
        sum_loss /= sample_num;
        const int64_t elapsed_sec = timer.ElapsedSeconds();
        const int64_t elapsed_min = elapsed_sec / 60 % 60;
        const int64_t elapsed_hou = elapsed_sec / 3600;
        std::stringstream ss;
        ss << std::setfill('0') << std::fixed;
        ss << std::setw(2) << elapsed_hou << ":";
        ss << std::setw(2) << elapsed_min << ":";
        ss << std::setw(2) << elapsed_sec % 60 << "\t";
        const float epoch_rate = 100.0f * (i + kBatchSize) / n_sample;
        ss << e << "\t";
        ss << step << "\t";
        ss << epoch_rate << "\t";
        ss << sum_loss << "\t";
        ofs << ss.str() << std::endl;
        std::cout << ss.str() << std::endl;
        sum_loss = 0;
        sample_num = 0;
      }

      optimizer.zero_grad();
      loss.backward();
      optimizer.step();
      torch::save(nerf, save_dir + "nerf_model.pt");
      if (step >= kMaxStep) {
        break;
      }
    }

    torch::save(nerf, save_dir + "nerf_model.pt");
    torch::save(optimizer, save_dir + "optimizer.pt");
    if (step >= kMaxStep) {
      break;
    }
  }
}
